src/ProvidedTypes.fs

```diff
@@ -8357,11 +8357,12 @@ namespace ProviderImplementation.ProvidedTypes
         member __.TranslateQuotationToCode qexprf (paramNames: string[]) (argExprs: Expr[]) =
             // Use the real variable names instead of indices, to improve output of Debug.fs
             // Add let bindings for arguments to ensure that arguments will be evaluated
-            let vars = argExprs |> Array.mapi (fun i e -> Var(paramNames.[i], e.Type))
+            let varDecisions = argExprs |> Array.mapi (fun i e -> match e with Var v when v.Name = paramNames.[i] -> false, v | _ -> true, Var(paramNames.[i], e.Type))

```

Comment:
This is breaking change @dsyme 
```
System.IndexOutOfRangeException
  HResult=0x80131508
  Message=Index was outside the bounds of the array.
  Source=SwaggerProvider.DesignTime
  StackTrace:
   at <StartupCode$SwaggerProvider-DesignTime>.$ProvidedTypes.varDecisions@8360.Invoke(Int32 i, FSharpExpr e) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 8360
   at Microsoft.FSharp.Collections.ArrayModule.MapIndexed[T,TResult](FSharpFunc`2 mapping, T[] array)
   at ProviderImplementation.ProvidedTypes.QuotationSimplifier.TranslateQuotationToCode(FSharpFunc`2 qexprf, String[] paramNames, FSharpExpr[] argExprs) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 8360
   at ProviderImplementation.ProvidedTypes.ProvidedTypesContext.convCodeToTgt[a](FSharpFunc`2 codeFun, Boolean isStatic, Boolean isCtor, ProvidedParameter[] parameters, Boolean isGenerated, FSharpList`1 argsT) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 8764
   at <StartupCode$SwaggerProvider-DesignTime>.$ProvidedTypes.xT@8877-20.Invoke(FSharpList`1 argsT) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 8877
   at ProviderImplementation.ProvidedTypes.ProvidedConstructor.GetInvokeCode(FSharpList`1 args) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 925
   at <StartupCode$SwaggerProvider-DesignTime>.$ProvidedTypes.Compile@14026-13.Invoke(ILTypeBuilder tb, FSharpOption`1 ptdT) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14087
   at <StartupCode$SwaggerProvider-DesignTime>.$ProvidedTypes.typeMembers@13867.Invoke(ProvidedTypeDefinition ptd) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 13868
   at ProviderImplementation.ProvidedTypes.AssemblyCompiler.iterateTypes[a](FSharpFunc`2 f, IEnumerable`1 providedTypeDefinitions) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 13880
   at <StartupCode$SwaggerProvider-DesignTime>.$ProvidedTypes.Compile@14026-12.Invoke(FSharpFunc`2 f, IEnumerable`1 providedTypeDefinitions) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14026
   at ProviderImplementation.ProvidedTypes.AssemblyCompiler.Compile(Boolean isHostedExecution) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14026
   at ProviderImplementation.ProvidedTypes.TypeProviderForNamespaces.Microsoft-FSharp-Core-CompilerServices-ITypeProvider-GetGeneratedAssemblyContents(Assembly assembly) in X:\SwaggerProvider\paket-files\fsprojects\FSharp.TypeProviders.SDK\src\ProvidedTypes.fs:line 14409
   at Microsoft.FSharp.Compiler.Tainted`1.Protect[a](FSharpFunc`2 f, range range)
```

<img width="1103" alt="2018-02-13_2316" src="https://user-images.githubusercontent.com/1197905/36171523-0a16c862-1114-11e8-89df-20ad840d0307.png">

Reply:
I do not fully understand how it worked before, but for generative types `argExprs.Length <> paramNames.Length`
In my case `paramNames` in the empty array and `argExprs.[0]` is `this`.

<img width="863" alt="2018-02-13_2337" src="https://user-images.githubusercontent.com/1197905/36172499-2147608e-1117-11e8-9114-b318328c48ff.png">

Reply:
Ugh, sorry. Will check

Reply:
@sergey-tihon Yeah, we had no tests for generative ProvidedConstructors, only erased ones.  Fixing this now.

Reply:
All fixed. Thanks for reporting so quickly. And the good thing that this is improving our test suite... https://github.com/fsprojects/FSharp.TypeProviders.SDK/pull/202